architecture-guide.md

```diff
@@ -0,0 +1,350 @@
+  for testing and storybook purposes, as in the example below.
+  ```qml
+  // Utils.qml (singleton)
+  
+  QtObject {
+    property var mainModuleInst: typeof mainModule !== "undefined" ? mainModule : null
+    property var sharedUrlsModuleInst: typeof  sharedUrlsModule !== "undefined" ? sharedUrlsModule : null
+    property var globalUtilsInst: typeof globalUtils !== "undefined" ? globalUtils : null
+  }
+  ```
+  The testing/storybook code also becomes complicated because it's important to
+  ensure that a singleton is mocked before the tested component is instantiated.
+
+  Stateless singleton also means not using components like `Settings` inside singletons.
+
+- Follow well known and valuable software programming principles, especially
+  SOLID (https://en.wikipedia.org/wiki/SOLID).
+
+  Well understood and applied `SOLID` principles lead to testable, flexible and
+  easily changeable code.
+  
+- The API of components should be well thought out. Expose dependencies,
+  hide internal details.
+
+  Ideally, it should be enough for a developer to read the public API of a given
+  component to understand what's needed to use it correctly.
+
+  - Examples of hidden dependencies:
+  - using stateful singletons - then components communicate to each other
+    implicitly, dependency is not visible in their public API
+  - accessing backend via singletons
+  - taking via public API much more than needed (e.g. component needs two
+    properties from store, but the whole store is provided with tens of
+    properties and methods). It's violation of `ISP` from `SOLID`.
+  - Examples of not hidden details:
+  - component contains searchable list and exposes search string assuming that
+    filtering will be done externally
+
+- Favor composition over parameterization and inheritance.
+
+  Overly parameterized components are fragile and not easy to maintain. Big set
+  of switches, altering the appearance and behavior of the component, introduces
+  a lot of conditional statements in the internal implementation. Finally,
+  it's difficult to assure that all combinations are valid or specify which ones
+  are valid and desired.
+
+- Action signals (e.g. click on delegate) should not provide metadata from the
+  model. Only `index` or unique `key` should be an argument. The necessary
+  metadata should be fetched on the signal receiver side.
+
+  ```qml
+  signal collectibleClicked(int chainId, string contractAddress, 
+                            string tokenId, string uid, int tokenType) // BAD
+
+  signal collectibleClicked(string key)                                // OK
+  ```
+  
+  Otherwise, the component unnecessarily bypasses some values from the model.
+  Callers may also want to access roles which are not used for displaying.
+  It leads to creating unnecessary requirements for the input model.
+
+- use consistently `key` as a unique identifier of the model, even if there are
```

Comment:
`id` itself is a bad name, QML reserved keyword

Reply:
Just for practical reason in `qml` environment. You cannot write:
```
ListElement {
  id: "my_id"
  name: "same name"
}
``` 
because `id` a special thing in qml.

I will add this explanation there.

Reply:
Indeed, I was thinking of properties only. Thanks